### ISSUE TITLE
Bad hornerE rewrite

### ISSUE THREAD

**Issue by SnarkBoojum** [math-comp/math-comp#920]
I thought `hornerE` would be able to solve that goal, but it couldn't:

```Coq
From mathcomp Require Import all_ssreflect all_algebra ssralg ssrint ssrnum.

From mathcomp.analysis Require Import reals.

From mathcomp.algebra_tactics Require Import ring.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.
Import Order.TTheory GRing.Theory.

Local Open Scope ring_scope.

Import GRing.

Section BadHornerE.

Variable RR: realType.

Definition RX := { poly RR }.
Definition Psum := (4 *: 'X^3 - 3 *: 'X + 1) : RX.
Variable a: RR.

Lemma nameless: Psum.[a] == 4 * a^+ 3 - 3 * a + 1.
Proof.
(* rewrite !hornerE. should be enough, I think, but gives an awful result *)
by rewrite !hornerD hornerN !hornerZ horner_exp hornerX hornerC.
Qed.

End BadHornerE.
```

**Comment by proux01** [comment]
@SnarkBoojum could you be more precise? `by rewrite !hornerE` works for me:
```Coq
From mathcomp Require Import all_ssreflect all_algebra.
From mathcomp.analysis Require Import reals.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Import GRing.Theory.

Local Open Scope ring_scope.

Section BadHornerE.

Variable RR: realType.
Variable a: RR.

Definition Psum : {poly RR} := 4 *: 'X^3 - 3 *: 'X + 1.

Lemma nameless: Psum.[a] == 4 * a^+ 3 - 3 * a + 1.
Proof. by rewrite !hornerE. Qed.

End BadHornerE.
```

Note (although this has no impact here) that the imports of `ssralg`, `ssrnum` and `ssrint` are superfluous as they are already included in `all_algebra`. It is also recommended *not* to import `GRing` as stated in the header of ssralg.v:
```Coq
(* The Lemmas about these structures are contained in both the GRing module   *)
(* and in the submodule GRing.Theory, which can be imported when unqualified  *)
(* access to the theory is needed (GRing.Theory also allows the unqualified   *)
(* use of additive, linear, Linear, etc). The main GRing module should NOT be *)
(* imported.                                                                  *)
```

**Comment by SnarkBoojum** [comment]
I was expecting the result of `rewrite !hornerE` to give me an equality of `4 * a ^ 3 - 3 * a + 1` with itself, but it gives a much more complex expression.

Of course, prefixing with `by` solves that simple goal ; but if you are inside a complete proof and just need a simplified expression for the polynomial evalutation, you get something awful, and I think that can be a problem.

**Comment by SnarkBoojum** [comment]
Why shouldn't `hornerE` itself preserve powers?

**Comment by SnarkBoojum** [comment]
Here is what the slightly changed hornerE I propose gives on my example:
```
From mathcomp Require Import all_ssreflect all_algebra.
From mathcomp.analysis Require Import reals.

Set Implicit Arguments.
Unset Strict Implicit.
Unset Printing Implicit Defensive.

Import GRing.Theory.

Local Open Scope ring_scope.

Section BadHornerE.

Variable RR: realType.
Variable a: RR.

Definition Psum : {poly RR} := 4 *: 'X^3 - 3 *: 'X + 1.

Lemma nameless: Psum.[a] == 4 * a^+ 3 - 3 * a + 1.
Proof.
 rewrite !hornerE. (* stop here and have a look at the expression *)
 by [].
Qed.

End BadHornerE.

Section MyHornerE.

Local Notation simp := Monoid.simpm.

Definition MyHornerE :=
  (hornerD, hornerN, hornerX, hornerC, horner_exp,
   hornerCM, hornerZ, hornerM, horner_cons).

Variable RR: realType.
Variable a: RR.

Definition Psum' : {poly RR} := 4 *: 'X^3 - 3 *: 'X + 1.

Lemma nameless': Psum'.[a] == 4 * a^+ 3 - 3 * a + 1.
Proof.
 rewrite !MyHornerE. (* stop here and have a look at the expression *)
 by [].
Qed.

End MyHornerE.
```

**Comment by affeldt-aist** [comment]
resolved by #923
